### ISSUE TITLE
Gradient Accumulation Question

### ISSUE THREAD

**Issue by chao1224** [apache/incubator-mxnet#6480]
I'm working on how MXNet accumulates gradient within one batch. By gradient accumulation, I mean for one batch update, the gradients for all data samples in this batch are accumulated by sum or mean. This should be done during `backward()`.

As I follow the code, the `GraphExecutor::Backward` will run operations through an engine. And to be more specific, in function `GraphExecutor::RunOps`, I found the gradient calculation node pushed into engine [here](https://github.com/dmlc/mxnet/blob/ecc3dfac418c581b307003f6ad4bab64c703a3b9/src/executor/graph_executor.cc#L804).

Then my question is, where specifically is the function to accumulate gradients during gradient calculation? I think this operation may register somewhere, but I didn't find it.

I haven't found too many helpful posts about this. Really appreciate it if someone can help.

**Comment by piiswrong** [comment]
Done by this function: https://github.com/dmlc/mxnet/blob/master/src/executor/graph_executor.cc#L90

**Comment by chao1224** [comment]
@piiswrong Exactly what I'm looking for, thanks.

**Comment by chao1224** [comment]
@piiswrong HI Junyuan, just have a follow up problem. I think the aggregation function is defined [here](https://github.com/dmlc/mxnet/blob/1974be4a345b4b048826652a63fe74960730736c/src/operator/tensor/elemwise_sum.cc#L21), do you know if there are some ways I can output the gradient of each data point in one batch online? Thank you in advance.

**Comment by chao1224** [comment]
I am able to output the exact gradients at each backward operator like in [softmax output](https://github.com/dmlc/mxnet/blob/master/src/operator/softmax_output-inl.h#L108) and [cnn v1](https://github.com/dmlc/mxnet/blob/master/src/operator/convolution_v1-inl.h#L180).

But I was hoping there are higher level functions I can output the gradients, like while we are doing element-wise-sum in AggregateGradient function.

**Comment by yajiedesign** [comment]
This issue is closed due to lack of activity in the last 90 days. Feel free to reopen if this is still an active issue. Thanks!

**Comment by Light--** [comment]
how to sum gradients of different shapes?
